python_modules/dagster/dagster/_cli/dev.py

```diff
@@ -326,3 +346,122 @@ def _workspace_opts_to_serialized_cli_args(workspace_opts: WorkspaceOpts) -> Seq
         args.append("--use-ssl")
 
     return args
+
+
+class DagsterDevCodeServerManager(AbstractContextManager):
+    """Context manager that manages the lifecycle of code servers launched by the dagster dev command."""
+
+    def __init__(
+        self,
+        instance: DagsterInstance,
+        workspace_load_target: Optional[WorkspaceLoadTarget],
+        code_server_log_level: str = "INFO",
+        server_command: GrpcServerCommand = GrpcServerCommand.API_GRPC,
+    ) -> None:
+        self._stack = ExitStack()
+
+        self._instance = check.inst_param(instance, "instance", DagsterInstance)
+        self._workspace_load_target = check.opt_inst_param(
+            workspace_load_target, "workspace_load_target", WorkspaceLoadTarget
+        )
+
+        self._grpc_server_registry: GrpcServerRegistry = self._stack.enter_context(
+            GrpcServerRegistry(
+                instance_ref=self._instance.get_ref(),
+                server_command=server_command,
+                heartbeat_ttl=WEBSERVER_GRPC_SERVER_HEARTBEAT_TTL,
+                startup_timeout=instance.code_server_process_startup_timeout,
+                log_level=code_server_log_level,
+                wait_for_processes_on_shutdown=instance.wait_for_local_code_server_processes_on_shutdown,
+                additional_timeout_msg=INCREASE_TIMEOUT_DAGSTER_YAML_MSG,
+            )
+        )
+
+        self._current_workspace: CurrentWorkspace = CurrentWorkspace(code_location_entries={})
+        # Is it ok that there's only one of these? When would this be something that needs to be called per location?
+        self.__shutdown_event = threading.Event()
+        self.__auto_restart_thread = threading.Thread(target=self.auto_restart_thread, daemon=True)
+        self.__auto_restart_thread.start()
+        self.__client_heartbeat_thread = threading.Thread(
+            target=self.client_heartbeat_thread, daemon=True
+        )
+        self.__client_heartbeat_thread.start()
+
+    def auto_restart_thread(self) -> None:
+        """Thread that automatically restarts the code server if it is not responding."""
+        while True:
+            self.__shutdown_event.wait(get_auto_restart_code_server_interval())
+            if self.__shutdown_event.is_set():
+                break
+            for process in self._grpc_server_registry.all_processes:
```

Comment:
i don't think all_processes is what you want here since it explicitly includes every process in the registry, even after it gets swapped out. (I don't think dagster dev actually does any swapping in practice, but still) 

could we instead add get_server_process that returns an Optional[GrpcServerProcess] (and uses the same lock that get_grpc_endpoint does?), returning the GrpcServerProcess if there is noe?